momepy/coins.py

```diff
@@ -0,0 +1,539 @@
+
+    Parameters
+    ----------
+    edge_gdf : GeoDataFrame
+        GeoDataFrame containing edge geometry of street network
+    angle_threshold : int, float (default 0)
+        the angle threshold for COINS algorithm. Segments will only be considered 
+        a part of the same street if deflection angle is above the threshold.
+
+    Returns
+    ----------
+    - series containing the stroke_group.
+    - new GeoDataFrame prior to performing merging. Refer to COINS paper for more details.
+    - new GeoDataFrame after merging. Refer to COINS paper for more details.
+
+    Examples
+    --------
+
+    >>> coins = momepy.COINS(streets)
+
+    >>> premerge = coins.premerge() 
+
+    >>> stroke_gdf = coins.stroke_gdf()  
+
+    >>> stroke_attr = coins.stroke_attribute() 
+    
+    """
+
+    def __init__(self, edge_gdf, angle_threshold=0):
+        self.edge_gdf = edge_gdf
+        self.gdfProjection = self.edge_gdf.crs
+        self.already_merged = False
+
+        # Get indices of original gdf
+        self.uv_index =  self.edge_gdf.index.tolist()
+
+        # Get line segments from edge gdf
+        self.lines = [list(value[1].coords) for value in edge_gdf['geometry'].iteritems()]
+
+        # split edges into line segments
+        self.splitLines()
+
+        # create unique_id for each individual line segment
+        self.uniqueID()
+
+        # Compute edge connectivity table
+        self.getLinks()
+
+        # Find best link at every point for both lines
+        self.bestLink()
+
+        # Cross check best links and enter angle threshold for connectivity
+        self.crossCheckLinks(angle_threshold)
+
+    def premerge(self):
+        return self.create_gdf_premerge()
+
+    def stroke_gdf(self):
+        if not self.already_merged: 
+            self.mergeLines()
+        return self.create_gdf_strokes()
+
+    def stroke_attribute(self):
+        if not self.already_merged:
+            self.mergeLines()
+        return self.add_gdf_stroke_attributes()
+
+    def splitLines(self):
+        outLine = []
+        self.tempArray = []
+        n = 0
+        #Iterate through the lines and split the edges
+        idx = 0
+        for line in self.lines:
+            for part in _listToPairs(line):
+                outLine.append([part, _computeOrientation(part), list(), list(), list(), list(), list(), list(), self.uv_index[idx]])
+                # Merge the coordinates as string, this will help in finding adjacent edges in the function below
+                self.tempArray.append([n, f'{part[0][0]}_{part[0][1]}', f'{part[1][0]}_{part[1][1]}'])
+                n += 1
+            idx += 1
+
+        self.split = outLine
+
+    def uniqueID(self):
+    #Loop through split lines, assign unique ID and
+    #store inside a list along with the connectivity dictionary
+        self.unique = dict(enumerate(self.split))
+
+    def getLinks(self):
+        global result
```

Comment:
This is something we don't need within a class. This line can go and all `results` can become `self.results`.